tree/ntuple/v7/inc/ROOT/RCluster.hxx

```diff
@@ -0,0 +1,183 @@
+/// \file ROOT/RCluster.hxx
+/// \ingroup NTuple ROOT7
+/// \author Jakob Blomer <jblomer@cern.ch>
+/// \date 2020-03-11
+/// \warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2020, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#include <ROOT/RNTupleUtil.hxx>
+
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#ifndef ROOT7_RCluster
+#define ROOT7_RCluster
+
+namespace ROOT {
+namespace Experimental {
+namespace Detail {
+
+
+// clang-format off
+/**
+\class ROOT::Experimental::ROnDiskPage
+\ingroup NTuple
+\brief A page as being stored on disk, that is packed and compressed
+
+Used by the cluster pool to cache pages from the physical storage. Such pages generally need to be
+uncompressed and unpacked before they can be used by RNTuple upper layers.
+*/
+// clang-format on
+class ROnDiskPage {
+private:
+   /// The memory location of the bytes
+   const void *fAddress = nullptr;
+   /// The compressed and packed size of the page
+   std::size_t fSize = 0;
+
+public:
+   /// On disk pages within a page source are identified by the column and page number. The key is used for
+   /// associative collections of on disk pages.
+   struct Key {
+      DescriptorId_t fColumnId;
+      NTupleSize_t fPageNo;
+      Key(DescriptorId_t columnId, NTupleSize_t pageNo) : fColumnId(columnId), fPageNo(pageNo) {}
+      bool operator ==(const Key &other) const {
+         return fColumnId == other.fColumnId && fPageNo == other.fPageNo;
+      }
+   };
+
+   ROnDiskPage() = default;
+   ROnDiskPage(void *address, std::size_t size) : fAddress(address), fSize(size) {}
+
+   const void *GetAddress() const { return fAddress; }
+   std::size_t GetSize() const { return fSize; }
+
+   bool IsNull() const { return fAddress == nullptr; }
+};
+
+} // namespace Detail
+} // namespace Experimental
+} // namespace ROOT
+
+// For hash maps ROnDiskPage::Key --> ROnDiskPage
+namespace std
+{
+   template <>
+   struct hash<ROOT::Experimental::Detail::ROnDiskPage::Key>
+   {
+      size_t operator()(const ROOT::Experimental::Detail::ROnDiskPage::Key &key) const
+      {
+         return ((std::hash<ROOT::Experimental::DescriptorId_t>()(key.fColumnId) ^
```

Comment:
Can you add text explaining why/if this is a good hash for the pair?

Reply:
Good point - I'm afraid I can't and probably it isn't very good.  We want something like `boost::hash_combine` or `llvm::hash_combine` in order to hash over multiple fields of a struct. @Axel-Naumann do you think it is worth adding `hash_combine` to the ROOT infrastructure code?

Reply:
For this round, this is fine as is.  The comment can be something like "quick and dirty hash, likely very sub-optimal, to be revised later."